x-pack/plugins/reporting/server/export_types/csv_searchsource/generate_csv/generate_csv.ts

```diff
@@ -293,7 +308,7 @@ export class CsvGenerator {
           // open a scroll cursor in Elasticsearch
           results = await this.scan(index, searchSource, settings);
           scrollId = results?._scroll_id;
-          if (results.hits?.total != null) {
+          if (results?.hits?.total != null) {
```

Comment:
Is the total hits accurate? I'm [looking at the documentation](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-your-data.html#track-total-hits) and seeing that it could be a lower bound and by default is limited to 10,000 records.

<img width="775" alt="image" src="https://user-images.githubusercontent.com/82822460/182566739-725288f8-ec02-401f-85bc-a44f605fd69e.png">

---

Also, even if it is accurate: if during the pagination process the number of documents in the index changes, what will be the outcome? Will we receive only the documents that were in the index when the pagination started? E.g. no new documents and, if some document was deleted, will we still receive it?

---

Curious, what determines if the total hits is a `long` or an object?

<img width="397" alt="image" src="https://user-images.githubusercontent.com/82822460/182565840-70616571-db5d-4b68-b158-4bcc5a9964fa.png">

Reply:
@vadimkibana these are really great observations! 

In regards to accuracy, my takeaway is we should be checking the value of `total.relation`, not just capturing `total.value`. However, the search documentation states:
> When set to true the search response will always track the number of hits that match the query accurately (e.g. `total.relation` will always be equal to "eq" when `track_total_hits` is set to `true`)."

Therefore, we shouldn't run into a state where `total.relation == 'gte'`.

But I think there are a couple of issues, still:
 - Reporting is implicitly depending on SearchSource to set the `track_total_hits: true` in the request. There is no explicit way for Reporting to ask that SearchSource does this
    - In my testing (adding console.log), I verified SearchSource is doing this. 
 - The code needs to handle the possibility that `hits.total` may be `estypes.SearchTotalHits` instead of a number.

Reply:
@vadimkibana you have brought up a lot of really good observations!

> I'm [looking at the documentation](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-your-data.html#track-total-hits) and seeing that it could be a lower bound and by default is limited to 10,000 records.

The search body we get from SearchSource is adding `track_total_hits: true`, which according to the documentation means that `total.relation` should always be `"eq"`.

Your insights prompted a few changes:
1. DEBUG logging of the `track_total_hits` value in the search body crafted by SearchSource
2. Handle the possibility that `total.hits` may be an object with `relation` and `value` keys
3. INFO logging for the `total.relation` value (or default to `"eq"` if `hits.total` is simply a number). 

See b07b4dfaf97

>  if during the pagination process the number of documents in the index changes, what will be the outcome?

It's an implementation detail of scan and scroll search in Elasticsearch. I assume there is a snapshot of the state preserved at the time the scroll context is created, but I'm not sure. We pass that scroll context with each followup request, which should keep the reference to that preserved state.

In the long-term, we should switch to using search sessions instead of scan and scroll, and that will bring an entirely new realm of implementation details to ask about.